core/src/main/java/org/infinispan/util/concurrent/BlockingHandler.java

```diff
@@ -0,0 +1,158 @@
+    * @param stage stage that may or may not be complete to apply
+    * @param biConsumer the blocking biConsumer
+    * @param traceId an identifier that can be used to tell in a trace when an operation moves between threads
+    * @param <V> stage value type
+    * @return a stage that is complete when the biConsumer is but retains the results from the original stage
+    */
+   <V> CompletionStage<V> whenCompleteBlocking(CompletionStage<V> stage,
+         BiConsumer<? super V, ? super Throwable> biConsumer, Object traceId);
+
+   /**
+    * When the provided stage is complete, continue the completion chain of the returned CompletionStage on the
+    * supplied executor. If tracing is enabled a trace message is printed using the object as an identifier to more
+    * easily track the transition between threads.
+    * <p>
+    * This method is useful when an asynchronous computation completes and you do not want to run further processing
+    * on the thread that returned it. An example may be that some blocking operation is performed on a special blocking
+    * thread pool. However when the blocking operation completes we will want to continue the processing of that result
+    * in a thread pool that is for computational tasks.
+    * <p>
+    * If the supplied stage is already completed when invoking this command, this will return an already completed
+    * stage, which means any additional dependent stages will run in the invoking thread.
+    * @param <V> return value type of the supplied stage
+    * @param delay the stage to delay the continuation until complete
+    * @param traceId the id to print when tracing is enabled
+    * @return a CompletionStage that when depended upon will run any callback in the supplied executor
+    */
+   <V> CompletionStage<V> continueOnNonBlockingThread(CompletionStage<V> delay, Object traceId);
+
+   /**
+    * Provided a publisher that is known to block, this will ensure that the publisher is subscribed on the
+    * blocking thread and any values published will be on a non blocking thread. Note that if a blocking thread
+    * subscribes to the publisher these threads will not be used and thus the entire Publisher is subscribed and
+    * observed on the invoking thread, unless the caller changes this.
+    * @param publisher the publisher that when subscribed to will block
+    * @param <V> the published entry types
+    * @return publisher that will not block the current thread
+    */
+   <V> Publisher<V> blockingPublisher(Publisher<V> publisher);
+
+   /**
+    * Provides a {@link BlockingExecutor} which is limited to the provided concurrency amount.
+    * @param name name of the limited blocking executor
+    * @param concurrency maximum amount of concurrent operations to be performed via the returned executor
+    * @return a blocking executor limited in the amount of concurrent invocations
+    */
+   BlockingExecutor limitedBlockingExecutor(String name, int concurrency);
```

Comment:
Feels kind of artificial, to keep using `BLOCKING_EXECUTOR` without injecting it and triggering the checkstyle error.

Reply:
Nope, not quite. The returned Executor resumes a given task on a non blocking thread, unless the task was submitted by a blocking thread.

Reply:
It feels artificial to me because we only need to replace the usage of `LimitedExecutor` to avoid the checkstyle error. It's also more "limited" than the original `LimitedExecutor`, which could run a task that does some blocking stuff and then some non-blocking stuff, releasing the permit only when the non-blocking stuff was done.

Reply:
> It feels artificial to me because we only need to replace the usage of `LimitedExecutor` to avoid the checkstyle error.

This doesn't just "avoid" the checkstyle. This ensures that the LimitedExecutor is used in a way to not leak the blocking threads after it is invoked. That is the entire point of anything with `BlockingManager` or any object produced from it.

> It's also more "limited" than the original LimitedExecutor, which could run a task that does some blocking stuff and then some non-blocking stuff, releasing the permit only when the non-blocking stuff was done.

This doesn't replace the `LimitedExecutor`. This is only to use the LimitedExecutor with blocking threads to ensure they don't escape.

And I am guessing you are referring to the Supplier invocation not consuming a permit but waiting on the stage does? There is no reason we can't change the `BlockingManager` over time if needed too :)